### PR TITLE
feat(informer-gen): add option WithCustomInformer 

### LINKED ISSUE #92792: Request to add informers.WithCustomListOptions to SharedInformerFactory
The k8s scheduler needs to watch a number of API objects such as Pods, Nodes, Services, ReplicaSet, PDBs, etc. So SharedInformerFactory is a perfect fit to watch/list objects among its codebase.

One problem we're facing is that we want to keep the ListOption as default for all objects except for Pods - for Pods, we want to exclude Pods whose `.status.phase` equals to Failed or Succeded. However, in current api-machinery code, there is only `informers.WithTweakListOptions`, which unfortunately applied the custom option to all listers. So does it make sense to add a `informers.WithCustomListOptions` to SharedInformerFactory? so that it's possible to apply a custom ListOptions for **PodInformer only**.
(BTW: an alternative is to add a "informers.WithCustomInformer" so caller is able to modify the sharedXYZInformer in place)

Ref: https://github.com/kubernetes/kubernetes/pull/92682#issuecomment-653734850

/kind feature
/sig api-machinery
/sig scheduling

### PR DESCRIPTION
**Pull request by knight42**
Signed-off-by: knight42 <anonymousknight96@gmail.com>

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test

/kind feature
> /kind flake

**What this PR does / why we need it**:

The k8s scheduler needs to watch running pods and now there are two PodInformers, a cutomized one in scheduler and the default one in `sharedInformerFactory`. The k8s scheduler hope to use only one PodInformer to eliminate unexpected bugs.

Ref: https://github.com/kubernetes/kubernetes/pull/92682#issuecomment-653734850

**Which issue(s) this PR fixes**:

Fixes #92792

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go [hunk 1]
```diff
@@ -142,6 +142,14 @@ func WithNamespace(namespace string) SharedInformerOption {
...
 }
 
+// WithCustomInformer sets a custom informer for the given informer type.
+func WithCustomInformer(obj {{.object|raw}}, informer {{.cacheSharedIndexInformer|raw}}) SharedInformerOption {
+	return func(factory *sharedInformerFactory) *sharedInformerFactory {
+		factory.informers[reflect.TypeOf(obj)] = informer
+		return factory
+	}
+}
+
```
Comment (knight42) [inline staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:152]: for reviewer: these lines are the major modification

File: staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/informers/externalversions/factory.go [hunk 2]
```diff
@@ -75,6 +75,14 @@ func WithNamespace(namespace string) SharedInformerOption {
...
 }
 
+// WithCustomInformer sets a custom informer for the given informer type.
+func WithCustomInformer(obj v1.Object, informer cache.SharedIndexInformer) SharedInformerOption {
+	return func(factory *sharedInformerFactory) *sharedInformerFactory {
+		factory.informers[reflect.TypeOf(obj)] = informer
+		return factory
+	}
+}
+
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/informers/externalversions/factory.go:85]: can't the same be done already today?

```
factory.InformerFor(obj, func(versioned.Interface, time.Duration) cache.SharedIndexInformer {
  return informer
})
```

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/informers/externalversions/factory.go:85]: Not that calling `InformerFor` is praticularly nice. I actually prefer the option func style.

Comment (knight42) [reply staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/informers/externalversions/factory.go:85]: @sttts you are right, I have overlooked that. so.. do you think it is worhwhile adding this option func?

Comment (Huang-Wei) [reply staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/informers/externalversions/factory.go:85]: Thanks @sttts. That perfectly fits my requirement.

I will leave it to you to decide whether to continue the "WithCustomInformer" idea or not.

**Comment by knight42** [comment]
close for inactivity
/close

**Pull request closed without merging**